torchdata/datapipes/iter/load/fsspec.py

```diff
@@ -0,0 +1,141 @@
+        masks: Union[str, List[str]] = "",
+    ) -> None:
+        _assert_fsspec()
+
+        self.root: str = root
+        self.masks = masks
+
+    def __iter__(self) -> Iterator[str]:
+        fs, path = fsspec.core.url_to_fs(self.root)
+        is_local = fs.protocol == "file" or not self.root.startswith(fs.protocol)
+        if fs.isfile(path):
+            yield self.root
+        else:
+            for file_name in fs.ls(path):
+                if not match_masks(file_name, self.masks):
+                    continue
+
+                # ensure the file name has the full fsspec protocol path
+                if file_name.startswith(fs.protocol):
+                    yield file_name
+                else:
+                    if is_local:
+                        abs_path = os.path.join(path, file_name)
+                    else:
+                        abs_path = posixpath.join(path, file_name)
+
+                    if self.root.startswith(fs.protocol):
+                        yield fs.protocol + "://" + abs_path
+                    else:
+                        yield abs_path
+
+
+@functional_datapipe("load_file_by_fsspec")
+class FSSpecFileLoaderIterDataPipe(IterDataPipe[Tuple[str, StreamWrapper]]):
+    r""":class:`FSSpecFileLoaderIterDataPipe`.
+
+    Iterable DataPipe to open files from input datapipe which contains fsspec paths
+    and yields a tuple of pathname and opened file stream.
+
+    Args:
+        source_datapipe: Iterable DataPipe that provides the pathnames or urls
+        mode: An optional string that specifies the mode in which the file is opened ('r' by default)
+    """
+
+    def __init__(self, source_datapipe: IterDataPipe[str], mode: str = "r") -> None:
+        _assert_fsspec()
+
+        self.source_datapipe: IterDataPipe[str] = source_datapipe
+        self.mode: str = mode
+
+    def __iter__(self) -> Iterator[Tuple[str, StreamWrapper]]:
+        for file_uri in self.source_datapipe:
+            fs, path = fsspec.core.url_to_fs(file_uri)
+            file = fs.open(path, self.mode)
+            yield file_uri, StreamWrapper(file)
+
+    def __len__(self) -> int:
+        return len(self.source_datapipe)
+
+
+@functional_datapipe("save_by_fsspec")
+class FSSpecSaverIterDataPipe(IterDataPipe[str]):
+    r"""
+    Iterable DataPipe that takes in a DataPipe of tuples of metadata and data, saves the data
+    to the target path (generated by the filepath_fn and metadata), and yields the resulting fsspec
+    path
+
+    Args:
+        source_datapipe: Iterable DataPipe with tuples of metadata and data
+        mode: Mode in which the file will be opened for write the data ("w" by default)
+        filepath_fn: Function that takes in metadata nad returns the target path of the new file
+    """
+
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe[Tuple[Any, U]],
+        mode: str = "w",
+        filepath_fn: Optional[Callable] = None,
+        *,
+        pathmgr=None,
```

Comment:
nit: you can remove this part. This is the specific API for `iopath`

Reply:
done